src/syntax/rules.rkt

```diff
@@ -2,48 +2,91 @@
 
 ;; Arithmetic identities for rewriting programs.
 
-(require "../common.rkt" "../errors.rkt" "types.rkt" "syntax.rkt" "sugar.rkt")
+(require "../common.rkt"
+         "../errors.rkt"
+         "sugar.rkt"
+         "syntax.rkt"
+         "types.rkt")
 
-(provide *rules* *simplify-rules* *fp-safe-simplify-rules* (struct-out rule))
+(provide *rules*
+         *simplify-rules*
+         *fp-safe-simplify-rules*
+         (struct-out rule))
 
 (module+ internals
-  (provide define-ruleset define-ruleset* register-ruleset! *rulesets*))
+  (provide define-ruleset
+           define-ruleset*
+           register-ruleset!
+           *rulesets*))
+
+;; A rule represents a "find-and-replace" pattern where `input` and `output`
+;; are patterns, `itypes` is a mapping from variable name to type
+;; (or representation) and `otype` is a type (or representation).
+(struct rule (name input output itypes otype)
+        #:methods gen:custom-write
+        [(define (write-proc rule port mode)
+           (fprintf port "#<rule ~a>" (rule-name rule)))])
 
-;; Rulesets
+;; A ruleset is a collection of rewrite rules sharing
+;;  - a type signature (variables have a type or representation)
+;;  - rule tags (support for certain tags may be toggled on or off)
+;; Beyond these restrictions, the grouping of rules is arbitrary.
+;; In practice, rules with the same theme are grouped into a ruleset.
+;;
+;; Herbie will apply all the rules listed below as well as those
+;; provided by plugins. Rules are applied regardless of which
+;; operations are actually supported by a platform since rewrite
+;; rules are intended to be mathematical.
 (define *rulesets* (make-parameter '()))
 
-(define (rule-ops-supported? rule)
-  (define (ops-in-expr expr)
-    (cond
-      [(list? expr)
-       (and (impl-exists? (car expr))
-            (for/and ([subexpr (cdr expr)])
-              (ops-in-expr subexpr)))]
-      [else true]))
-  (ops-in-expr (rule-output rule)))
+;; Ruleset contract
+(define ruleset? (list/c (listof rule?) (listof symbol?) dict?))
 
+;; Updates the `*ruleset* parameter
+(define/contract (add-ruleset! name ruleset)
+  (-> symbol? ruleset? void?)
+  (when (dict-has-key? (*rulesets*) name)
+    (warn 'rulesets "Duplicate ruleset ~a, skipping" name))
+  (*rulesets* (dict-set (*rulesets*) name ruleset)))
+
+;; Rules: fp-safe-simplify ⊂ simplify ⊂ all
 ;;
-;; Rules
-;; fp-safe-simplify ⊂ simplify ⊂ all
-;;
-;; all                    requires at least one tag of an active group of rules
-;; simplify               same req. as all + 'simplify' tag
-;; fp-safe-simplify       same req. as simplify + 'fp-safe' tag ('fp-safe' does not imply 'simplify')
-;;
+;; all - at least one tag of an active group of rules
+;; simplify - subset of `all` that has the `simplify` tag
+;; fp-safe-simplify - subset of `simplify` that has `fp-safe` tag
+
+(define (ops-in-expr expr)
```

Comment:
Could this be something like `(remove-duplicates (map first (filter list? (all-subexpressions expr))))`? Maybe we gotta remove `if` also?

Reply:
Yep

Reply:
Oh, you know, I was thinking of some other `all-subexpressions` we have in the codebase but now that I think of it they are subtly different. Worth unifying sometime but let's not hold up the PR. This is good stuff and we should merge.